syntax = "proto3";

option java_package = "pman";
option java_outer_classname = "PmanProto";

package pman;

service Pman {
  rpc ProgramStatus (StatusRequest) returns (stream ProgramStatusReply) {}
  rpc StartProgram (StartRequest) returns (stream ProgramStatusReply) {}
  rpc StopProgram (StopRequest) returns (stream ProgramStatusReply) {}
  rpc RestartProgram (RestartRequest) returns (stream ProgramStatusReply) {}
}

message StatusRequest {
  string name = 1;
}

message StartRequest {
  string name = 1;
}

message StopRequest {
  string name = 1;
}

message RestartRequest {
  string name = 1;
}

message ProgramStatusReply {
  enum State {
    STOPPING = 0;
    RUNNING = 1;
  }
  enum ErrorStatus {
    OK = 0;
    NO_PROGRAMS = 1;
    NO_SUCH_PROGRAM = 2;
    ALREADY_STOPPING = 3;
    ALREADY_RUNNING = 4;
  }
  message Error {
    ErrorStatus status = 1;
    string message = 2;
  }

  string name = 1;
  State state = 2;
  Error error = 3;
}
